pkg/actor/validate_version.go

```diff
@@ -106,65 +116,81 @@ func (v *versionChecker) Act(ctx context.Context, cluster *resource.Cluster) err
 		log.Error(err, "cannot create k8s client")
 		return errors.Wrapf(err, "check version failed to create kubernetes clientset")
 	}
+	//we get the job to see it already exists
 	if err := v.client.Get(ctx, key, job); err != nil {
-		err := WaitUntilJobPodIsRunning(ctx, clientset, job, log)
-		if err != nil {
-			log.Error(err, "job not found")
+		//if job does not exist we create it
+		if err = reconcileJob(cluster, v.scheme, r, jobName, log); err != nil {
 			return err
 		}
+	} else {
+		// if job exists check if image changed
+		if dbContainer, err := kube.FindContainer(resource.JobContainerName, &job.Spec.Template.Spec); err != nil {
+			// this will fix the reconcile on job on immutable prop, somehow the job probably terminates
+			log.Error(err, "the job already exists but no container was found")
+			return nil
+		} else if dbContainer.Image != cluster.GetCockroachDBImageName() {
+			log.V(DEBUGLEVEL).Info("HERE 2", "job", jobName, "dbContainer.Image", dbContainer.Image, "spec image", cluster.GetCockroachDBImageName())
+			if err = reconcileJob(cluster, v.scheme, r, jobName, log); err != nil {
+				return err
+			}
+		}
+	}
+	// we comment this because after reconcile the object it is modified in k8s and
+	// this will trigger a restart.
+	// TODO: we need to delete this lines
+	// if changed {
+	// 	log.V(DEBUGLEVEL).Info("created/updated job, stopping request processing")
+	// 	CancelLoop(ctx)
+	// 	return nil
+	// }
+	log.V(DEBUGLEVEL).Info("version checker", "job", jobName)
+
+	// sometimes the Get on the job will return an error and an empty job struct.
+	// this is especially valid for kind cluster in our e2e
+	// To fix this we wait for the pod of the job to run and than get the job.
+	if err := waitUntilJobPodIsRunning(ctx, clientset, jobName, job.Namespace, r.Labels.Selector(), log); err != nil {
+		log.Error(err, "job pod not found")
+		return err
+	}
+
+	//we get the job after we checked that the pod is running
+	if err := v.client.Get(ctx, key, job); err != nil {
+		msg := fmt.Sprintf("failure: retrieved job%+v", *job)
+		log.Error(err, msg)
+		return err
 	}
 
-	// check if the job is completed or failed before EXEC
+	// check if the job is completed or failed before getting the logs from the pod
 	if finished, _ := isJobCompletedOrFailed(job); !finished {
-		if err := WaitUntilJobPodIsRunning(ctx, clientset, job, v.log); err != nil {
+		//we check first to see that the job pod it is running
+		if err := waitUntilJobPodIsRunning(ctx, clientset, jobName, job.Namespace, r.Labels.Selector(), v.log); err != nil {
 			// if after 2 minutes the job pod is not ready and container status is ImagePullBackoff
 			// We need to stop requeueing until further changes on the CR
 			image := cluster.GetCockroachDBImageName()
-			if errBackoff := IsContainerStatusImagePullBackoff(ctx, clientset, job, log, image); errBackoff != nil {
+			if errBackoff := isContainerStatusImagePullBackoff(ctx, clientset, job, log, image); errBackoff != nil {
```

Comment:
@chrislovecnm this is the part that confuses me, we return `nil` from `isContainerStatusImagePullBackoff` if the pod is in a ready state. Which is surprising from what the function name implies.